Sources/xcodeproj/Objects/Files/PBXGroup.swift

```diff
@@ -160,7 +160,15 @@ public extension PBXGroup {
         let groupPath = try fullPath(sourceRoot: sourceRoot)
 
         if let existingFileReference = try projectObjects.fileReferences.first(where: {
-            try filePath == $0.value.fullPath(sourceRoot: sourceRoot)
+            // Optimization: compare lastComponent before fullPath compare
+            guard let fileRefPath = $0.value.path else {
+                return try filePath == $0.value.fullPath(sourceRoot: sourceRoot)
+            }
+            let fileRefLPC = (fileRefPath as NSString).lastPathComponent
```

Comment:
I will change in force unwrap, because I try to follow the general style and rules of this repository.
But I would like to convey my point of view.

I believe that using force unwrap is a bad style for a user product (the auxiliary library is also intended for the user (another developer)). There are many articles, discussions and linter rules to prove this (for example: Swiftlint have this type of rule). Of course there is an alternative point of view.

I'll tell you why this is bad for me as a user of this library. I use this library as a dependency of my utility. My utility is a command line tool. If something went wrong, I want the user to get a distinct error. Therefore, I carefully check all parameters and extreme cases, pick up an error and display it. Most of the time I use exceptions for this. Since it provides a convenient mechanism for forwarding errors through the call stack. And they also allow you to add folbek, if this is an expected problem (ideally, of course, there should be no such problems). By the same principle, system methods work, for example methods of working with the file system.

When a force failure happens in some kind of dependency, I cannot catch it, I cannot display an error to the user. He will only see that the work was interrupted without giving a reason.
If this were an exception, then I would catch this situation and report that some part of the work could not be done, because your project file may be corrupted.

Reply:
@CognitiveDisson that's a totally valid point of view. I agree with you that being explicit by raising errors helps the consumer of the library catch those errors and therefore show a nicer message to the users: "hey! this unexpected thing happened?"

However, I think depending of the type of error it should be treated differently. I can think of two scenarios when the last path component of the path can be nil:
- When there's a bug on xcodeproj's implementation.
- When something has be wrongly configured in the Xcode project.

If it's the first scenario, I don't think it's a good idea to expose the consumer of the API to those errors. We should instead have tests that can catch implementation bugs. If it's the second scenario, we shouldn't design an API assuming that the user might not know what they are doing. We can design the API to prevent a bad usage, but there's some responsibility that lays on the consumer's side.

I think Core Data is a good example of this. I've seen many developers complaining about Core Data not being thread safe and that resulting in their apps crashing. Core Data needs to be understood to be used properly. Try to use Core Data in unsafe ways... you'll come across assertions that crash your apps. Should they then expose those as errors instead? I don't think so.

Having said that, there's some work on our side to be done:
- We need more documentation that expalins users how things work.
- There's still room for improvement on making the API safer.